# Tarea 1 

Alumnos:

* Miguel Espinoza
* Sebastián López
* Pedro Pérez

## P1

### Parte 1

In [1]:
import os

folder = 'data/raw/'  #Carpeta principal en la que estan los archivos
weeks = os.listdir(folder)  #Se agregan las subcarpetas a una lista
weeks

['w13', 'w14', 'w15', 'w16', 'w17']

In [2]:
#Se crean dos listas para guardar los archivos normales
#y los 'furnished'
files = []
files_furnished = []

#Mediante un bucle for se guardan las direcciones de todos
#los archivos en las listas creadas
for week in weeks:
    direccion = folder + week + '/'
    file = os.listdir(direccion)
    files.append(direccion + file[0])
    files_furnished.append(direccion + file[1])
    
files

['data/raw/w13/metrocuadrado_all_w13.csv',
 'data/raw/w14/metrocuadrado_all_w14.csv',
 'data/raw/w15/metrocuadrado_all_w15.csv',
 'data/raw/w16/metrocuadrado_all_w16.csv',
 'data/raw/w17/metrocuadrado_all_w17.csv']

In [3]:
import pandas as pd

#Se lee el primer dataframe para observar los datos
data_all = pd.read_csv(files[0])
data_all.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

En primer lugar se combinan los dataframes normales

In [4]:
data_all = pd.read_csv(files[0])

for file in files:
    dato_semanal = pd.read_csv(file)
    data_all = data_all.merge(dato_semanal, how = 'outer') 
    #Se pone 'outer' para hacer la union de todos los dataframes

print('Cantidad de datos: ', len(data_all))
data_all.head(3)

Cantidad de datos:  16470


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

In [5]:
#Se ve si hay duplicados
print('Duplicados: ', data_all.duplicated().sum())

Duplicados:  175


In [6]:
#Se eliminan las filas duplicadas
data_all.drop_duplicates(inplace=True)
print('Cantidad de datos: ', len(data_all))

Cantidad de datos:  16295


Ahora se combinan los archivos furnished

In [7]:
data_furnished = pd.read_csv(files_furnished[0])

for file in files_furnished:
    dato_semanal = pd.read_csv(file)
    data_furnished = data_furnished.merge(dato_semanal, how='outer')

print('Cantidad de datos: ', len(data_furnished))
data_furnished.head(3)

Cantidad de datos:  2101


property_type|rent_type|location       price  n_rooms  \
0  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000      3.0   
1  Casa en Venta Y Arriendo, BELLAVISTA Bogotá D.C..  $1.300.000      3.0   
2  Casa en Arriendo, PORTALES DEL NORTE Bogotá D.C..  $1.800.000      3.0   

   n_bath   surface                                            details  \
0     2.0  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
1     2.0   72.0 m2  hermosa casa en venta y en arriendo en, mandal...   
2     3.0  76.81 m2  Estrato 4. CASA TRIPLEX!! Exterior . Cuenta co...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
1  https://www.metrocuadrado.com/inmueble/venta-c...                 41.4  
2  https://www.metrocuadrado.com/inmueble/arriend...                 41.3

In [8]:
#Se buscan los duplicados y se eliminan
print('Duplicados: ', data_furnished.duplicated().sum())

Duplicados:  2


In [9]:
data_furnished.drop_duplicates(inplace=True)

print('Cantidad de datos: ', len(data_furnished))
#Se observa que son muchos menos los datos de tipo furnished

Cantidad de datos:  2099


Al intentar combinar ambos dataframes usando `data_all.merge(data_furnished, how='outer')` aparece el error:

`ValueError: You are trying to merge on object and float64 columns. If you wish to proceed you should use pd.concat`

Por lo que se decide comprobar los tipos de las columnas

In [10]:
#Se observan las columnas de los dataframe
print('Tipos Data All')
print(data_all.dtypes)
print()
print('Tipos Data Furnished')
print(data_furnished.dtypes)

Tipos Data All
property_type|rent_type|location     object
price                                object
n_rooms                              object
n_bath                               object
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object

Tipos Data Furnished
property_type|rent_type|location     object
price                                object
n_rooms                             float64
n_bath                              float64
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object


Se observa que el error se debe a que en `Data All` las columnas `n_rooms` y `n_bath` son de tipo `object`, en cambio en `Data Furnished` son de tipo `float64`.

Este problema se resuelve simplemente cambiando el tipo de esas columnas.

In [11]:
#Se cambian las columnas de 'data_furnished' a object
data_furnished = data_furnished.astype({'n_rooms':'object', 'n_bath':'object'})

#Obs: No se pueden cambiar las columnas de 'data_all' a float64, pues hay un dato llamado '5+'

In [153]:
#Se combinan los dataframes
data = data_all.merge(data_furnished, how='outer', indicator='Furnished')

#Se coloca utiliza el parámetro indicator para indicar de que dataframe 
#proviene cada fila.
#left_only indica que viene del dataframe 'data_all'
#right_only indica que viene del dataframe 'data_furnished'
#y both que proviene de ambos

print('Cantidad de datos: ', len(data))
data.head(3)

Cantidad de datos:  18394


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

   Furnished  
0  left_only  
1  left_only  
2  left_only

In [154]:
data['Furnished'].replace({'left_only':'No', 'right_only':'Si', 'both':'Ambos'}, inplace=True)

#Se cambian los nombres para indicar más claramente si provienen de furnished o no
data.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished  
0        No  
1        No  
2        No

In [155]:
#Se cuentan la cantidad de cada valores de cada tipo
data['Furnished'].value_counts()

No    16295
Si     2099
Name: Furnished, dtype: int64

Como se observa no hay elementos comunes entre ambos grupos, son disjuntos. No hay archivos con texto 'furnished' que no estén contenidos en archivos con texto 'all'.

In [156]:
#Esto se comprueba tambien haciendo una intersección entre los dataframes
#mediante el argumento 'inner'
pd.merge(data_all, data_furnished, how='inner')

#Se observa que el resultado es un dataframe vacío

Empty DataFrame
Columns: [property_type|rent_type|location, price, n_rooms, n_bath, surface, details, url, metrocuadrado_index]
Index: []

### Parte 2

#### A

Primero se limpia el precio

In [157]:
data['price'].head(3)

0    $1.050.000
1    $3.930.000
2      $750.000
Name: price, dtype: object

Se observa que son de tipo `object`, pues el precio contiene el signo peso '$' y los puntos. Convendría eliminar este signo y los puntos para poder pasar los datos a `float64`.

In [158]:
data['price'] = data['price'].str.replace('$','')
data['price'] = data['price'].str.replace('.','')
data['price'].head(3)

0    1050000
1    3930000
2     750000
Name: price, dtype: object

In [159]:
#Se cambia el tipo a float64
data['price'] = data['price'].astype('float')
data['price'].head(3)

0    1050000.0
1    3930000.0
2     750000.0
Name: price, dtype: float64

Ahora se limpia el área

In [160]:
data['surface'].tail(4)

18390    164.34 m2
18391     110.0 m2
18392     106.0 m2
18393     135.0 m2
Name: surface, dtype: object

Se deben eliminar los carácteres ' m2' al final del número, y luego cambiar el tipo a `float`.

In [161]:
data['surface'] = data['surface'].str.replace(' m2','')
data['surface'] = data['surface'].astype('float')
data['surface'].tail(4)

18390    164.34
18391    110.00
18392    106.00
18393    135.00
Name: surface, dtype: float64

Ahora se limpia el número de habitaciones

In [162]:
data['n_rooms'].head(3)

0    3
1    3
2    3
Name: n_rooms, dtype: object

Parece que fueran números enteros, pero el tipo de la columna es object

In [163]:
#Se ven los valores de la columna
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, '5+', 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

Se observa que la columna tiene valores de tipo string y float. Más aún se ve que se podría cambiar inmediatamente el tipo a `float` sino existiera el valor `5+`.

In [164]:
#Se cuenta cuantas entradas hay de cada valor
data['n_rooms'].value_counts()

2      6571
1      6271
3      1351
4      1198
1.0    1040
2.0     938
5       862
3.0      52
5.0      36
4.0      32
5+        9
Name: n_rooms, dtype: int64

Se ve que solo son 9 las entradas con valor `5+`, por lo que es una parte infima del total del dataset que es de $18394$ entradas, simplemente se podrían eliminar. Pero como no se desea botar datos, se cambiará el valor `5+` por $6$, dado que es el entero mayor a 5 más cercano.

In [165]:
data['n_rooms'] = data['n_rooms'].replace('5+',6)
#Se reemplazan el valor que daba problemas, y se muestran los valores de las entradas
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, 6, 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

In [166]:
#Como esta correcto, se puede cambiar el tipo de la columna
data['n_rooms'] = data['n_rooms'].astype('float')
data['n_rooms'].head(3)

0    3.0
1    3.0
2    3.0
Name: n_rooms, dtype: float64

Por último se limpia el número de baños

In [167]:
data['n_bath'].tail(3)

18391    3
18392    3
18393    2
Name: n_bath, dtype: object

Nuevamente parecen ser enteros, pero el tipo es object

In [168]:
#Se ven los valores de la columna
data['n_bath'].unique()

array(['2', '1', '4', '3', nan, '5', '5+', 2.0, 3.0, 4.0, 5.0, 1.0],
      dtype=object)

Nuevamente el error viene dado por `5+`, por lo que se procede de igual manera que en `n_rooms`.

In [169]:
data['n_bath'] = data['n_bath'].replace('5+',6) #Se reemplazan '5+' por 6
data['n_bath'] = data['n_bath'].astype('float') #Se cambia el tipo a float
data['n_bath'].head(3)

0    2.0
1    2.0
2    1.0
Name: n_bath, dtype: float64

Finalmente se muestran nuevamente los tipos de las columnas

In [170]:
data.dtypes

property_type|rent_type|location     object
price                               float64
n_rooms                             float64
n_bath                              float64
surface                             float64
details                              object
url                                  object
metrocuadrado_index                 float64
Furnished                            object
dtype: object

#### B

Se debe dividir la columna `property_type|rent_type|location` en 3 columnas distintas.

In [171]:
#Se hace una copia de la data por si se realiza un error
copia_seguridad = data.copy()

In [151]:
#Cuando algo salga mal correr esta celda
data = copia_seguridad.copy()

In [172]:
#Se muestran unos pocos datos de esta columna
data['property_type|rent_type|location']

0          Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..
1        Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...
2        Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...
3              Casa en Arriendo, PARIS GAITAN Bogotá D.C..
4        Casa en Arriendo, BALCONES DE ORIENTE Bogotá D...
                               ...                        
18389    Apartamento en Arriendo, CHICO NORTE Bogotá D.C..
18390    Apartamento en Arriendo, CHICO NORTE II Bogotá...
18391    Apartamento en Venta Y Arriendo, CHICO NORTE B...
18392    Apartamento en Arriendo, MOLINOS NORTE Bogotá ...
18393       Apartamento en Arriendo, EL NOGAL Bogotá D.C..
Name: property_type|rent_type|location, Length: 18394, dtype: object

Se pueden deducir varias cosas:
1. El tipo de propiedad es la primera palabra del string.
2. El tipo de renta aparece después de la palabra 'en' y un espacio ' '.
3. El lugar esta después de la coma.

En primer lugar se decide dividir el string en dos partes separadas por la coma.

In [173]:
new_columns = data['property_type|rent_type|location'].str.split(pat=',', expand=True)
new_columns.head(3)

0                                           1
0  Casa en Arriendo               LA SOLEDAD NORTE Bogotá D.C..
1  Casa en Arriendo    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..
2  Casa en Arriendo   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

In [174]:
#Se insertan las columnas en la data 
data[['property_type|rent_type', 'location']] = new_columns

#Se elimina la columna original
data.drop('property_type|rent_type|location', axis=1, inplace=True)

data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type  \
0        No        Casa en Arriendo   
1        No        Casa en Arriendo   
2        No        Casa en Arriendo   

                                     location  
0               LA SOLEDAD NORTE Bogotá D.C..  
1    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..  
2   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

Ahora se divide la columna `property_type|rent_type` en dos columnas más.

In [175]:
new_columns = data['property_type|rent_type'].str.split(pat='en ', expand=True)
new_columns.head(3)

0         1
0  Casa   Arriendo
1  Casa   Arriendo
2  Casa   Arriendo

In [176]:
#Se insertan las columnas en la data y se elimina la columna original
data[['property','rent']] = new_columns
data.drop('property_type|rent_type', axis=1, inplace=True)
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                                    location property      rent  
0        No               LA SOLEDAD NORTE Bogotá D.C..    Casa   Arriendo  
1        No    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..    Casa   Arriendo  
2        No   VILLA DE LOS ALPES II SECTOR Bogotá D.C..    Casa   Arriendo

Por último se revisa que no hayan espacios vacíos al inicio o final de las nuevas columnas.

In [178]:
print(data['location'][0])
print(data['property'][0])
print(data['rent'][0])

 LA SOLEDAD NORTE Bogotá D.C..
Casa 
Arriendo


Tanto location como property tienen espacios, por lo que se utiliza strip para eliminarlos.

In [182]:
data['location'] = data['location'].str.strip()
data['property'] = data['property'].str.strip()

### Parte 3

### Parte 4

### Parte 5

### Parte 6

## P2

# P3 

## Implementación

### Parte 1

In [213]:
from sklearn.base import BaseEstimator as BE
from sklearn.base import RegressorMixin as RM
import numpy as np

class RegresionBayesianaEmpirica(BE,RM):
    
    def __init__(self,alpha,beta,tol=1e-15,maxiter=100):
        self.set_params(alpha=alpha,beta=beta,tol=tol,maxiter=maxiter)
        
    def get_posteriori(self,X,y,alpha,beta):
        N,d = X.shape
        try:
            S_N = np.linalg.inv(alpha*np.eye(d) + beta*np.dot(X.T,X)) # dim dxd
        except LinAlgError:
            print('Matriz S_N no invertible')
        m_N = beta*np.dot(S_N,np.dot(X.T,y)) # dim dx1

        return m_N, S_N
    
    
    def fit(self,X,y,i=0,verbose=True):
        i += 1
        if verbose:
            print('Iteracion',i)
            print('Valores actuales: alpha=',self.alpha,' beta=',self.beta)
        if i <= self.maxiter:
            N,d = X.shape
            eigen_val, eigen_vec = np.linalg.eig(self.beta*np.dot(X.T,X))
            eigen_val = np.real_if_close(eigen_val,tol=1)
            gamma = np.sum(eigen_val/(self.alpha + eigen_val))
            m_N, S_N = self.get_posteriori(X,y,self.alpha,self.beta)
        
            new_alpha = gamma/(np.dot(m_N.T,m_N))
            new_beta = (N-gamma)/(np.sum((y-np.dot(m_N.T,X.T))**2))
        
            if np.abs(self.alpha - new_alpha) < self.tol and np.abs(self.beta - new_beta) < self.tol:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                if verbose:
                    print('Valores finales: alpha=',self.alpha,' beta=',self.beta)
            else:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                self.fit(X,y,i)
                
    
    def predict(self,X_,return_std=False):
        y_ = np.dot(X_,self.m_N)
        y_std = np.sqrt((1/self.beta) + np.diag(np.dot(X_,np.dot(self.S_N,X_.T))))
        
        if return_std:
            return (y_,y_std)
        return y_

### Parte 2

In [214]:
import pandas as pd
import numpy as np

#Datos para prueba
df = pd.read_csv('data_2.csv')

In [215]:
df.replace([np.inf, -np.inf], np.nan, inplace=True)
df.dropna(inplace=True)
df.drop(df[df.surface==0].index, inplace=True)
df.drop(['price','surface','url','details','personas','property'],axis=1,inplace=True)#

In [216]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15773 entries, 3 to 18018
Data columns (total 19 columns):
n_rooms                   15773 non-null float64
n_bath                    15773 non-null float64
metrocuadrado_index       15773 non-null float64
Garajes                   15773 non-null int64
tipo de producto          15773 non-null int64
trabajo_ninos             15773 non-null float64
trabajoamp_ninos          15773 non-null float64
jovenes_nini              15773 non-null float64
indice_envegecimiento     15773 non-null float64
jefe_mujer                15773 non-null float64
adultos_mayores_pobres    15773 non-null float64
indice_inseguridad        15773 non-null float64
areas_verdes_perc         15773 non-null float64
pob_density               15773 non-null float64
Precio m2                 15773 non-null float64
location                  15773 non-null object
rent                      15773 non-null object
Furnished                 15773 non-null object
UPZRecat        

In [217]:
df_y = df['Precio m2']#.to_numpy()
df_X = df.drop('Precio m2',axis=1)#.to_numpy()

#num_cols = df_X.select_dtypes(include=['int64', 'float64']).columns
#cat_cols = df_X.select_dtypes(include=['object']).columns
float_cols = ['pob_density']
percentage_cols = ['n_rooms','n_bath','metrocuadrado_index','Garajes','trabajo_ninos','trabajoamp_ninos',
                   'jovenes_nini','indice_envegecimiento','jefe_mujer','adultos_mayores_pobres','indice_inseguridad',
                   'areas_verdes_perc']
cat_cols = ['Furnished','location','rent','tipo de producto','UPZRecat']

In [218]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

#cat pipe
cat_pipe = Pipeline(steps=[('ordinal', OneHotEncoder(handle_unknown='ignore'))])

#num pipe
#num_pipe = Pipeline(steps=[('poly',PolynomialFeatures(degree=3)),('scaler',StandardScaler())])
float_pipe = Pipeline(steps=[('scaler',StandardScaler()),('poly',PolynomialFeatures(degree=3))])
percentage_pipe = Pipeline(steps=[('scaler',MinMaxScaler()),('poly',PolynomialFeatures(degree=3))])

#Se diferencian columnas numéricas y categóricas
prep = ColumnTransformer(transformers=[('flt',float_pipe,float_cols),('perc',percentage_pipe,percentage_cols),('cat',cat_pipe,cat_cols)])

### Parte 3

In [219]:
random_state = 7
def run(df_X,df_y,mode='all'):
    from sklearn.model_selection import train_test_split
    import numpy as np
    import pickle
    
    alpha_grid = np.linspace(1e-10,1e-5,3,endpoint=True)
    beta_grid = np.linspace(1e-10,1e-5,3,endpoint=True)
    
    actual_R2 = 0
    actual_RMSE = 0
    
    X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
    def run_single(pipeline):
        pipeline.fit(X_train,y_train)
        print('Fit terminado', '\n')
        y_pred = pipeline.predict(X_test) 

        R_2 = pipeline.score(X_test, y_test)
        RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
        return R_2, RMSE
    
    best_alpha = alpha_grid[0]
    best_beta = beta_grid[0]
    for alpha in alpha_grid:
        for beta in beta_grid:
            Tubo = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(alpha,beta))])
            print("Empezando fit: alpha y beta iniciales",[alpha,beta])
            R_2, RMSE = run_single(Tubo)
            if R_2 > actual_R2:
                actual_R2 = R_2
                actual_RMSE = RMSE
                best_alpha = alpha
                best_beta = beta
    
    best_fit = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(best_alpha,best_beta))])
    print("Reentrenando el mejor modelo encontrado")
    best_fit.fit(X_train,y_train)
    if mode=='all':
        filename = 'modelo.pkl'
        pickle.dump(best_fit,open(filename,'wb'))
    elif mode=='EDA':
        filename = 'modelo_EDA.pkl'
        pickle.dump(best_fit,open(filename,'wb'))
    else:
        raise InvalidMode('Modo de ejecución no válido')
    
    print('RMSE:',actual_RMSE,' R_2:',actual_R2)

In [220]:
run(df_X,df_y)

Empezando fit: alpha y beta iniciales [1e-10, 1e-10]
Iteracion 1
Valores actuales: alpha= 1e-10  beta= 1e-10
Iteracion 2
Valores actuales: alpha= 3.8290427430317627e-08  beta= 1.9628446636610682e-08
Iteracion 3
Valores actuales: alpha= 5.247242747453211e-08  beta= 1.9546591661525524e-08
Iteracion 4
Valores actuales: alpha= 5.975830318299633e-08  beta= 1.9499364217949887e-08
Iteracion 5
Valores actuales: alpha= 6.280197087432148e-08  beta= 1.9477840885152242e-08
Iteracion 6
Valores actuales: alpha= 6.396641567401383e-08  beta= 1.9469257009780898e-08
Iteracion 7
Valores actuales: alpha= 6.439708170090721e-08  beta= 1.9466028688623654e-08
Iteracion 8
Valores actuales: alpha= 6.455437767669053e-08  beta= 1.9464842112764012e-08
Iteracion 9
Valores actuales: alpha= 6.461156572344813e-08  beta= 1.9464409706787802e-08
Iteracion 10
Valores actuales: alpha= 6.463232296171194e-08  beta= 1.946425262580393e-08
Iteracion 11
Valores actuales: alpha= 6.463985255151921e-08  beta= 1.9464195627895717e-08

Iteracion 2
Valores actuales: alpha= 3.829042743040566e-08  beta= 1.9628446636609868e-08
Iteracion 3
Valores actuales: alpha= 5.247242747464667e-08  beta= 1.954659166152471e-08
Iteracion 4
Valores actuales: alpha= 5.975830318309763e-08  beta= 1.9499364217949153e-08
Iteracion 5
Valores actuales: alpha= 6.280197087430695e-08  beta= 1.9477840885152364e-08
Iteracion 6
Valores actuales: alpha= 6.396641567389849e-08  beta= 1.946925700978177e-08
Iteracion 7
Valores actuales: alpha= 6.439708170083688e-08  beta= 1.9466028688624187e-08
Iteracion 8
Valores actuales: alpha= 6.45543776767666e-08  beta= 1.9464842112763437e-08
Iteracion 9
Valores actuales: alpha= 6.461156572333664e-08  beta= 1.9464409706788646e-08
Iteracion 10
Valores actuales: alpha= 6.463232296158232e-08  beta= 1.946425262580491e-08
Iteracion 11
Valores actuales: alpha= 6.463985255148791e-08  beta= 1.9464195627895956e-08
Iteracion 12
Valores actuales: alpha= 6.464258327568605e-08  beta= 1.946417495440098e-08
Iteracion 13
Valores ac

Iteracion 12
Valores actuales: alpha= 6.464258327563541e-08  beta= 1.946417495440137e-08
Iteracion 13
Valores actuales: alpha= 6.46435735369262e-08  beta= 1.946416745712629e-08
Iteracion 14
Valores actuales: alpha= 6.464393263176585e-08  beta= 1.9464164738376875e-08
Iteracion 15
Valores actuales: alpha= 6.464406284773063e-08  beta= 1.9464163752491187e-08
Iteracion 16
Valores actuales: alpha= 6.464411006678923e-08  beta= 1.9464163394987528e-08
Iteracion 17
Valores actuales: alpha= 6.46441271894607e-08  beta= 1.9464163265348728e-08
Iteracion 18
Valores actuales: alpha= 6.464413339840142e-08  beta= 1.9464163218339723e-08
Iteracion 19
Valores actuales: alpha= 6.46441356499349e-08  beta= 1.9464163201292953e-08
Valores finales: alpha= 6.464413646638456e-08  beta= 1.946416319511147e-08
Fit terminado 

Reentrenando el mejor modelo encontrado
Iteracion 1
Valores actuales: alpha= 1e-10  beta= 1e-05
Iteracion 2
Valores actuales: alpha= 4.527801176934891e-11  beta= 1.9931133205165777e-08
Iteracion

### Parte 4

In [223]:
#Ahora, seleccionamos solo aquellas columnas que se obtienen del análisis EDA
df_y = df['Precio m2']#.to_numpy()
df_X = df[['n_rooms','metrocuadrado_index','tipo de producto','indice_envegecimiento','jefe_mujer',
                  'Furnished','UPZRecat']]#.to_numpy()

#num_cols = df_X.select_dtypes(include=['int64', 'float64']).columns
#cat_cols = df_X.select_dtypes(include=['object']).columns
float_cols = []
percentage_cols = ['n_rooms','metrocuadrado_index','indice_envegecimiento','jefe_mujer']
cat_cols = ['Furnished','tipo de producto','UPZRecat']

In [226]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

#cat pipe
cat_pipe = Pipeline(steps=[('ordinal', OneHotEncoder(handle_unknown='ignore'))])

#num pipe
#num_pipe = Pipeline(steps=[('poly',PolynomialFeatures(degree=3)),('scaler',StandardScaler())])
float_pipe = Pipeline(steps=[('scaler',StandardScaler()),('poly',PolynomialFeatures(degree=3))])
percentage_pipe = Pipeline(steps=[('scaler',MinMaxScaler()),('poly',PolynomialFeatures(degree=3))])

#Se diferencian columnas numéricas y categóricas
prep = ColumnTransformer(transformers=[('flt',float_pipe,float_cols),('perc',percentage_pipe,percentage_cols),('cat',cat_pipe,cat_cols)])

In [227]:
run(df_X,df_y,mode='EDA')

Empezando fit: alpha y beta iniciales [1e-10, 1e-10]
Iteracion 1
Valores actuales: alpha= 1e-10  beta= 1e-10
Iteracion 2
Valores actuales: alpha= 8.111795241418061e-09  beta= 1.522523912398418e-08
Iteracion 3
Valores actuales: alpha= 5.058518036005735e-09  beta= 1.5264585137037273e-08
Iteracion 4
Valores actuales: alpha= 3.5662055039023164e-09  beta= 1.5291348590848967e-08
Iteracion 5
Valores actuales: alpha= 2.8155825871450023e-09  beta= 1.530796101590647e-08
Iteracion 6
Valores actuales: alpha= 2.4350804297152657e-09  beta= 1.5317326273691068e-08
Iteracion 7
Valores actuales: alpha= 2.241464420026334e-09  beta= 1.5322308671756167e-08
Iteracion 8
Valores actuales: alpha= 2.142689884334832e-09  beta= 1.5324892732171605e-08
Iteracion 9
Valores actuales: alpha= 2.0922156474346977e-09  beta= 1.532622092082614e-08
Iteracion 10
Valores actuales: alpha= 2.066397985059938e-09  beta= 1.5326901728496497e-08
Iteracion 11
Valores actuales: alpha= 2.0531851283735516e-09  beta= 1.5327250435812625e-

Iteracion 23
Valores actuales: alpha= 2.0393438162659733e-09  beta= 1.5327615895320932e-08
Iteracion 24
Valores actuales: alpha= 2.0393340802709747e-09  beta= 1.5327616152437472e-08
Iteracion 25
Valores actuales: alpha= 2.0393290937307964e-09  beta= 1.5327616284126346e-08
Iteracion 26
Valores actuales: alpha= 2.0393265418011222e-09  beta= 1.5327616351519924e-08
Iteracion 27
Valores actuales: alpha= 2.039325233318529e-09  beta= 1.5327616386075477e-08
Valores finales: alpha= 2.0393245632820616e-09  beta= 1.5327616403770382e-08
Fit terminado 

Empezando fit: alpha y beta iniciales [5.0000500000000005e-06, 5.0000500000000005e-06]
Iteracion 1
Valores actuales: alpha= 5.0000500000000005e-06  beta= 5.0000500000000005e-06
Iteracion 2
Valores actuales: alpha= 8.111795241800731e-09  beta= 1.522523912398031e-08
Iteracion 3
Valores actuales: alpha= 5.058518035813044e-09  beta= 1.5264585137039996e-08
Iteracion 4
Valores actuales: alpha= 3.5662055042212508e-09  beta= 1.529134859084277e-08
Iteracion 

Iteracion 8
Valores actuales: alpha= 2.2082804177296175e-09  beta= 1.5323174192319122e-08
Iteracion 9
Valores actuales: alpha= 2.1257396098672895e-09  beta= 1.532533828835217e-08
Iteracion 10
Valores actuales: alpha= 2.083547540434728e-09  beta= 1.5326449405707323e-08
Iteracion 11
Valores actuales: alpha= 2.0619624064004937e-09  beta= 1.5327018770956208e-08
Iteracion 12
Valores actuales: alpha= 2.050914606101217e-09  beta= 1.5327310374832218e-08
Iteracion 13
Valores actuales: alpha= 2.0452587177525483e-09  beta= 1.532745970237663e-08
Iteracion 14
Valores actuales: alpha= 2.0423628386225927e-09  beta= 1.5327536169676513e-08
Iteracion 15
Valores actuales: alpha= 2.040880021317467e-09  beta= 1.532757532674343e-08
Iteracion 16
Valores actuales: alpha= 2.0401207283428346e-09  beta= 1.532759537817495e-08
Iteracion 17
Valores actuales: alpha= 2.0397319175402554e-09  beta= 1.532760564606295e-08
Iteracion 18
Valores actuales: alpha= 2.039532817557711e-09  beta= 1.5327610904025322e-08
Iteracion 

### Parte 5

In [228]:
from sklearn.linear_model import BayesianRidge
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
BR_pipe = Pipeline([('prep',prep),('clf',BayesianRidge())])
BR_pipe.fit(X_train,y_train)
print('Fit terminado', '\n')
y_pred = BR_pipe.predict(X_test) 

R_2 = BR_pipe.score(X_test, y_test)
RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
print('RMSE = ',RMSE,' R2 =',R_2,)

Fit terminado 

RMSE =  8083.34595688417  R2 = 0.5905168918323402
